### ISSUE TITLE
Query (get) helper

### ISSUE THREAD

**Issue by ErisDS** [TryGhost/Ghost#4439]
One of the main new features we need to add to Ghost is a 'query helper' for themes. This is a helper that gives theme developers access to the API to make custom queries for data they'd like to display.

This helper will only provide access to the `GET` API endpoints that don't require user authentication - that is getting published posts, and all tags, with all the sorting and filtering options which are available. 

I propose that this helper be a block helper called 'fetch':

```
{{#fetch}}
  .. do stuff with the data here
{{/fetch}}
```

After that, there are a number of ways we could specify the details of the API call we want to make:

1. Just accept a string parameter that contains the API call URL:
`{{#fetch "/posts/?limit=5"}}`
`{{#fetch "/posts/12/?include=tags"}}`

2. Take the resource as a parameter, and use named attributes for the query params:
`{{#fetch "posts" limit="5"}}`
`{{#fetch "posts" id="12" include="tags"}}`

3. Also have the resource be a named attribute:

`{{#fetch resource="posts" limit="5"}}`
`{{#fetch resource="posts" id="12" include="tags"}}`

4. Some other format?

We need to find a format that makes sense to theme developers. Using the API call URL as a string is very simple to implement, but all the `/`'s and `?`'s and `&`'s look quite complex even though URLs are fairly well understood thing.

The second option is kind of inconsistent with itself, but it looks nice. It's also similar to doing something like `{{date published_at format="dd mm YY"}}` where there can be a parameter and an attribute, and just like the date helper defaults to published_at, the fetch helper can default to posts.

The third one makes everything consistent, but as soon as you start using named parameters you need to do something like `id="12"` or `slug="welcome-to-ghost"` if you want to specify an id, or a slug etc to fetch a specific resource.

Thoughts on a postcard please!

**Comment by ErisDS** [comment]
The more I think about it, the more I prefer the second option as well:

`{{#fetch "posts" limit="5"}}`
`{{#fetch "posts" id="12" include="tags"}}`

However, I'm leaning towards using `get` instead of fetch?

`{{#get "posts" limit="5"}}`
`{{#get "posts" id="12" include="tags"}}`

**Comment by dbalders** [comment]
Personally, I like #get over #fetch and I like option 2 as well. Seems the most straight forward and readable.

**Comment by ErisDS** [comment]
Here's a little bit of how I imagine it working:

In post.hbs, display the 5 most recent posts in a sidebar:

```
{{#post}}
<article>
  <header><h1>{{title}}</h1></header>
  <section>{{content}}</section>
  <footer>Published at {{date format="DD MMMM YYYY"}}</footer>
</article>
{{/post}}
{{#get "posts" limit="5"}}
  {{#if posts}}
    <aside>
      <ol>
      {{#foreach posts}}
        <li>{{title}}</lI>
      {{/foreach}}
      </ol>
    </aside>
  {{/if}}
{{/get}}
```

In index.hbs, display a list of all tagsin a sidebar:

```
{{#foreach posts}}
    {{! output partials/post.hbs}}
    {{> "post"}}
{{/foreach}}
{{#get "tags" limit="all" order="alpha"}}
  {{#if tags}}
    <aside>
      <ul>
      {{#foreach tags}}
        <li>{{name}}</li>
      {{/foreach}}
    </ul>
  </aside>
  {{/if}}
{{/get}}
```

**Comment by halfdan** [comment]
Instead of the `{{#if resource}}` wouldn't it make sense to not execute the block if the query result is empty?

**Comment by ErisDS** [comment]
@halfdan good catch!! Updated my comment to demonstrate this + else.

**Comment by ErisDS** [comment]
An update on the status of this issue, which also applies to #4799. 

I began work on implementing this, and ran into a *small* oversight - express-hbs's fancy async helpers don't work for block helpers. Essentially, the options don't get passed around the same way so access to `fn` and `inverse` are lost. I started investigating the whys and what-fors to see if this could easily be resolved in express-hbs or is better suited to a custom workaround in Ghost and haven't come to a conclusion just yet. Once this is fixed the implementation of `{{#get}}` is straightforward.

Note: Handlebars itself has [no intention of supporting async helpers](https://github.com/wycats/handlebars.js/issues/717).

**Comment by ofstudio** [comment]
My current workaround with redefined `registerAsyncHelper` function
https://gist.github.com/ofstudio/e13276a9c587ecd41a95

**Comment by ErisDS** [comment]
@ofstudio thanks for dropping by and sharing that gist with us - it gave me some inspiration and now the issue is fixed in [express-hbs 0.8.3](https://github.com/barc/express-hbs/releases/tag/v0.8.3) ;)

**Comment by gaurav21r** [comment]
@ErisDS +1 :+1: For this feature! Certainly theme developers are running into a wall without this. 

2 Questions though:

1. Expected time of readiness (without testing and all)?
2. Would be great if we could see work on this issue in a branch or a fork? I would like to contribute to this effort but see no PRs or branches linking to this issue?

**Comment by ErisDS** [comment]
This helper is relatively simple, the problem at the moment is the API itself is still very, very limited so even if I put it in master, the helper wouldn't do enough things. As it has (so far) not been accessible except for in the admin interface, it's got a few gnarly bits and limitations which are far too tailored to that scenario and not any where near general enough.

There's a balance to be had here, because I know as soon as the `{{#get}}` helper becomes available the limitations will become obvious and more people will likely be interested in fixing them. The flip side is that if we make the helper available and then make major changes to the API, themes will break everywhere. So the balance is to make a few major changes up front:

The most important one of those is the horrid `staticPages` parameter, which I just raised an issue for:  #5151, which itself is blocked on #5093. If we could get these things smoothed out, I think it'd be OK to ship a `{{#get}}` helper with lots of warnings that it's beta only - and see how it goes.

Some other notes on API limitations:
- there is no way to affect the order of posts, tags or users. There's an [open issue](https://github.com/TryGhost/Ghost/issues/889) for adding the ability to do reverse-chronological posts, but it's one of the oldest issues we have open! - issue #889
- there's no way to fetch posts filtered by the featured flag yet, although I think someone is working on that - issue #5152
- the tag & user endpoint will let you fetch a single resource, a page of resources, or all resources - there are no filters available at all (as well as no orders)

So I imagine the second we put a helper out there, theme devs are going to want more options :grin:

**Comment by PDXIII** [comment]
+1 
I want this feature so badly!!!
Do think it will be available in 0.6.5 even as beta with a lot of deep red warnings?

**Comment by ErisDS** [comment]
Sorry, this issue totally deserves an update. 

I've been working on it for quite a while. Now that I've fixed `express-hbs` to allow for block ajax helpers, the helper itself is a relatively simple PR. However, all the work that is needed to make it useful is quite extensive.

Firstly, the API needs to support 'public access' permissions, permitting access to published posts and active users, whilst keeping drafts etc private. I've been working on this in PR #5496, but there is more work to do so there will be an issue for this soon. 

Secondly the API needs to do useful things like filtering and sorting. I explained many of these in my [previous comment](https://github.com/TryGhost/Ghost/issues/4439#issuecomment-93789838) but not a great deal of progress has been made. I recently opened a [new issue](https://github.com/TryGhost/Ghost/issues/5463) describing my plan for how to implement this and am planning to break this down into smaller issues over the coming days.

Thirdly, there are considerations for the design of the helper itself - like how to allow for passing local data in so that it is possible to do dynamic queries. E.g, I'm on a post and I want to list the most recent 5 posts which have the same tags (a rough version of related posts). That means making the following work:

```
{{#post}}
{{#get "posts" limit="5" tag=post.tags}}{{/get}}
{{/post}}
```

Or following the new [parameter design](https://github.com/TryGhost/Ghost/issues/5463), making something like this work:

```
{{#post}}
{{#get "posts" limit="5" filter="tag:{{post.tags}}" }}{{/get}}
{{/post}}
```

In summary, there's quite a lot to do, but not really for this particular issue. The plan is to provide a 'beta' of the get helper in the next release.

**Comment by PDXIII** [comment]
Totally dope!

**Comment by ErisDS** [comment]
I'm still working away on this, looking to put together an initial WIP PR which covers this + the filter parameter in the next day or so, such that both things can be tested out and played with in a useful setting.

I thought I'd update with a few things that have come up as a result of putting the get helper together. 

## Path Expressions and JSONPath

The first is the obvious need to be able to pass in data from the local context to form queries, as mentioned in my previous comment. Doing this with the filter query stuff in place in a rudimentary form, I've managed to get the following working:

```
{{#post}}
  {{#get "posts" limit="5" filter="tag:{{post.tags}}" }}
     <ol>
    {{#foreach posts}}
      <li><a href="{{url}}">{{title}}</a></li>
    {{/foreach}}
    </ol>
  {{/get}}
{{/post}}
```

The handlebars data expressions inside the filter query are being parsed out by simply looking for matched pairs of `{{` and `}}` and JSONPath is being used to take the path between them and fetch the correct piece of data from the context. This works surprisingly well :) It should be possible to do the same thing for other parameters, as well as the `filter` parameter.

## Block params

Handlebars has a new feature for helpers called block params. I found out about this when working on the `{{#foreach}}` helper, essentially it allows you to rename the data coming out of the helper so....

`{{#get "posts" filter="featured:true" as |featured|}}` lets you fetch posts which are featured as an array of objects called `featured`. The result is:

```
{{#get "posts" filter="featured:true" as |featured| }}
  <ol>
  {{#foreach featured}}
    <li><a href="{{url}}">{{title}}</a></li>
  {{/foreach}}
  </ol>
{{/get}}
```

I think that's the key points!

**Comment by lulato** [comment]
Any news on this issue. I'm currently working on a theme that will take advantage of the GET feature, so I was wondering if I should work around it or wait.

**Comment by ErisDS** [comment]
If you look at the links above your comment, you'll see there are currently 2 open PRs against this issue - they are the news :wink:. Rather than working around or waiting, I recommend testing them out and letting us know how you get on - feedback will help us ship.

**Comment by lulato** [comment]
:confused: 
Fine I'll do it, but just cause you seem nice. And you have like 230 day strike on Github :astonished:

**Comment by gazpachu** [comment]
Hi @ErisDS I want to help with the testing of this feature. Do I have to wait for the WIP to be merged into master or can I check it out already? Thanks

**Comment by kevinansfield** [comment]
@gazpachu You can checkout PRs for local testing. There's a good guide to make the process super simple on the dev blog - http://dev.ghost.org/easy-git-pr-test/

**Comment by gazpachu** [comment]
Thanks @kevinansfield. I was able to checkout the PR and test it. I've placed the get helper in the post template (like in ErisDS example). It returns all the posts correctly, but the filter doesn't seem to be working. I've tried to filter only the 'getting-started' tag but it still renders links to posts without that tag. The error handling doesn't render any error message either. I've also noticed in the admin area that the tags field is not present, so I can't add tags to new posts from the admin. Is that expected? Are you adding tags manually using an external DB GUI? What else can I test?

**Comment by ErisDS** [comment]
Hi @gazpachu, #5496 is already merged and so is included whenever you check out one of the get helper PRs. If you want to play with the more advanced filter features then I recommend checking out #5608, rather than #5619. 

The tag field for adding tags to posts is now under the post settings menu, but you'll need the latest copy of the PRs as it was missing entirely for a while.

I have updated both of the PRs - #5619 is just the get helper & #5608 is the get helper & filter feature - they have both been rebased on master today so when you check one of them out you'll have all of the code as of today + these extra features, which should make it easier to test.

**Comment by sodbileg** [comment]
Hello. How is this going? I'm currently developing a theme for Envato Ghost Theme Contest and would like to know the status.

**Comment by Ba5nanas** [comment]
i need offset of post

**Comment by ThienNguyenMinh** [comment]
Hi,
I am deverloper. I am really looking forward to this feature.

**Comment by DanielTamkin** [comment]
@ThienNguyenMinh :+1:

**Comment by dnascimento** [comment]
+1 waiting looking forward to this feature. @ErisDS I agree with your syntax. I'm looking for: 
{{#get "tags" limit="all" order="alpha"}}
Please, let me know if further testing is needed. Since you are working on it, do you have any updated branch with that feature? I can test and give feedback.

**Comment by oswaldoacauan** [comment]
:+1:

**Comment by Alucardology** [comment]
I'm just wondering when this feature will be available? 
Thanks in advance.

**Comment by ErisDS** [comment]
It is coming in Beta in the next release. That's the only thing I can say for certain :wink:

**Comment by DanielTamkin** [comment]
I'm So damn excited for this feature @ErisDS !

**Comment by online** [comment]
Hi!

Someone seen theme that supports the new query `{{#get}}` to display all tags in the sidebar or elsewhere? I use a standard theme Casper.

**Comment by dbalders** [comment]
@online the beta version of the get helper just came out so i doubt any themes already have it running. You can however check out how to do it yourself here (http://themes.ghost.org/v0.7.2/docs/get#section-make-a-tag-cloud) or here (https://www.ghostforbeginners.com/how-to-add-a-tag-menu-to-your-blog/)

**Comment by online** [comment]
@dbalders thank you, this is just what I need! :wink:

**Comment by krokofant** [comment]
I've looked a bit through the roadmap and I'm still wondering when the get helper will be able to get static pages. I want to be able to get a page via the slug of the page just like you can with regular posts.

Any insights? :)
